src/workerd/api/encoding.c++

```diff
@@ -460,29 +461,204 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
+          position += encodeUtf8CodeUnit(c, out + position);
+          pendingSurrogate = false;
+        }
+      }
+    } else if (isLeadSurrogate(c)) {
+      pendingSurrogate = true;
+    } else {
+      if (isTrailSurrogate(c)) {
+        position += encodeUtf8CodeUnit(0xFFFD, out + position);
+      } else {
+        position += encodeUtf8CodeUnit(c, out + position);
+      }
+    }
+  }
+
+  if (pendingSurrogate) {
+    encodeUtf8CodeUnit(0xFFFD, out + position);
+  }
+}
+
 }  // namespace
 
-jsg::BufferSource TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
-  auto str = input.orDefault(js.str());
-  auto view = JSG_REQUIRE_NONNULL(jsg::BufferSource::tryAlloc(js, str.utf8Length(js)), RangeError,
-      "Cannot allocate space for TextEncoder.encode");
-  [[maybe_unused]] auto result = encodeIntoImpl(js, str, view);
-  KJ_DASSERT(result.written == view.size());
-  return kj::mv(view);
+jsg::Ref<TextEncoder> TextEncoder::constructor(jsg::Lock& js) {
+  return js.alloc<TextEncoder>();
+}
+
+jsg::JsUint8Array TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
+  jsg::JsString str = input.orDefault(js.str());
+  std::shared_ptr<v8::BackingStore> backingStore;
+  size_t utf8_length = 0;
+
+  // Fast path: check if string is one-byte before creating ValueView
+  if (str.isOneByte(js)) {
+    auto length = str.length(js);
+    // Allocate buffer for Latin-1. Use v8::ArrayBuffer::NewBackingStore to avoid creating
+    // JS objects during conversion.
+    backingStore = v8::ArrayBuffer::NewBackingStore(
+        js.v8Isolate, length, v8::BackingStoreInitializationMode::kUninitialized);
+    auto backingData = reinterpret_cast<kj::byte*>(backingStore->Data());
+
+    str.writeOneByte(js, kj::ArrayPtr<kj::byte>(backingData, length),
+        jsg::JsString::WriteFlags::REPLACE_INVALID_UTF8);
+
+    utf8_length =
+        simdutf::utf8_length_from_latin1(reinterpret_cast<const char*>(backingData), length);
+
+    if (utf8_length == length) {
+      // ASCII fast path: no conversion needed, Latin-1 is same as UTF-8 for ASCII
+      auto array = v8::Uint8Array::New(v8::ArrayBuffer::New(js.v8Isolate, backingStore), 0, length);
+      return jsg::JsUint8Array(array);
+    }
+
+    // Need to convert Latin-1 to UTF-8
+    std::shared_ptr<v8::BackingStore> backingStore2 = v8::ArrayBuffer::NewBackingStore(
+        js.v8Isolate, utf8_length, v8::BackingStoreInitializationMode::kUninitialized);
+    auto written = simdutf::convert_latin1_to_utf8(reinterpret_cast<const char*>(backingData),
+        length, reinterpret_cast<char*>(backingStore2->Data()));
+    KJ_DASSERT(utf8_length == written);
+    auto array =
+        v8::Uint8Array::New(v8::ArrayBuffer::New(js.v8Isolate, backingStore2), 0, utf8_length);
+    return jsg::JsUint8Array(array);
+  }
+
+  // Two-byte string path
+  {
+    // Note that ValueView flattens the string, if it's not already flattened
+    v8::String::ValueView view(js.v8Isolate, str);
```

Comment:
Given that `ValueView` forces flattening, which we need to avoid... let's instead use `WriteV2` to write to an off-heap allocation (e.g. `kj::SmallArray<uint16_t, 4096> buffer(length)`) to avoid the additional GC pressure caused by flattening.